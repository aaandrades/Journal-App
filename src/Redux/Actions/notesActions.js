import swal from "sweetalert";
import { db } from "../../firebase/firebase-config";
import { loadNotes } from "../../helpers/loadNotes";
import { uploadFile } from "../../helpers/uploadFile";
import { types } from "../Types/types";

export const createNoteAction = () => {
  return async (dispatch, getState) => {
    try {
      const { uid } = getState().auth;
      const newNote = { title: "", body: "", date: new Date().getTime() };
      const doc = await db.collection(`${uid}/journal/notes`).add(newNote);

      dispatch(activeNoteAction(doc.id, newNote));
      dispatch(setNoteAction({ id: doc.id, ...newNote }));
    } catch (error) {
      console.error(error);
      swal("Error creating note", error, "error");
    }
  };
};

export const activeNoteAction = (id, note) => ({
  type: types.noteActive,
  payload: { ...note, id },
});

export const updateNoteAction = (note) => ({
  type: types.notesUpdated,
  payload: note,
});

export const fetchNotes = (uid) => {
  return async (dispatch) => {
    const notes = await loadNotes(uid);
    dispatch(setNotesAction(notes));
  };
};

export const setNotesAction = (notes) => ({
  type: types.notesLoad,
  payload: notes,
});

export const setNoteAction = (note) => ({
  type: types.noteCreated,
  payload: note,
});

export const startSaveNoteAction = (note) => {
  return async (dispatch, getState) => {
    try {
      const { uid } = getState().auth;
      const buildNote = { ...note };
      delete buildNote.id;
      await db.doc(`/${uid}/journal/notes/${note.id}`).update(buildNote);
      dispatch(refreshNotes(note.id, buildNote));
      swal("Saved", note.title, "success");
    } catch (error) {
      console.error(error);
      swal("Error saving :(", error, "error");
    }
  };
};

export const refreshNotes = (id, note) => ({
  type: types.notesUpdatedState,
  payload: {
    id,
    note: { id, ...note },
  },
});

export const closeNoteAction = () => ({
  type: types.emptyNote,
});

export const startUploading = (file) => {
  return async (dispatch, getState) => {
    const { active: activeNote } = getState().notes;
    swal({
      title: "Uploading...",
      text: "We're uploading your image, hold on tight",
      icon: "info",
      closeOnClickOutside: false,
      buttons: false,
    });

    const url = await uploadFile(file);
    activeNote.url = url;
    dispatch(startSaveNoteAction(activeNote));
    swal.close();
  };
};

export const startDeletingAction = (id) => {
  return async (dispatch, getState) => {
    try {
      const { uid } = getState().auth;
      swal({
        title: "Deleting...",
        text: "We're deleting your note, hold on tight",
        icon: "info",
        closeOnClickOutside: false,
        buttons: false,
      });
      await db.doc(`${uid}/journal/notes/${id}`).delete();
      dispatch(deleteNoteAction(id));
      swal.close();
    } catch (error) {
      console.error("Error: ", error);
      swal("Error saving :(", error, "error");
    }
  };
};

export const deleteNoteAction = (id) => ({
  type: types.notesDelete,
  payload: id,
});

export const cleanNotesAction = () => ({
  type: types.notesLogoutCleaning,
});
