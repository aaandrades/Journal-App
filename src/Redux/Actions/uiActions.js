import { types } from "../Types/types";

export const setErrorAction = (message) => ({
  type: types.uiSetError,
  payload: message,
});

export const setRemoveErrorAction = () => ({
  type: types.uiRemoveError,
});

export const enableLoadingAction = () => ({
  type: types.enableLoading,
});

export const disableLoadingAction = () => ({
  type: types.disableLoading,
});
