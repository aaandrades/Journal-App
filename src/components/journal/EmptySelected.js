import React from "react";

export const EmptySelected = () => {
  return (
    <div className="empty__main-content">
      <p>Select something</p>
      <i className="far fa-star fa-4x mt-5"></i>
    </div>
  );
};
