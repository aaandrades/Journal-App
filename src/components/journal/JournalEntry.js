import React from "react";
import { dateOrdinal } from "../../helpers/helpers";

export const JournalEntry = ({ id, title, body, date, url, handleClick }) => {
  const getMonth = new Date(date).toLocaleString("en-US", {
    weekday: "long",
  });

  const getDay = new Date(date).getDay();

  return (
    <div
      className="journal__entry pointer animate__animated animate__backInLeft animate__faster"
      onClick={() => handleClick(id, { title, body, date, id, url })}
    >
      {url && (
        <div
          className="journal__entry-picture"
          style={{
            backgroundSize: "cover",
            backgroundImage: `url(${url})`,
          }}
        ></div>
      )}
      <div className="journal__entry-body">
        <p className="journal__entry-title">{title}</p>
        <p className="journal__entry-content">{body}</p>
      </div>
      <div className="journal__entry-date">
        <span>{getMonth}</span>
        <h4>{dateOrdinal(getDay)}</h4>
      </div>
    </div>
  );
};
