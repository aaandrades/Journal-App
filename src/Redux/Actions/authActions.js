import { firebase, googleAuthProvider } from "../../firebase/firebase-config";
import { types } from "../Types/types";
import { disableLoadingAction, enableLoadingAction } from "./uiActions";
import { cleanNotesAction } from "./notesActions";
import swal from "sweetalert2";

// Login with email and password
export const initLoginAction = (email, password) => {
  return async (dispatch) => {
    try {
      dispatch(enableLoadingAction());
      const { user } = await firebase
        .auth()
        .signInWithEmailAndPassword(email, password);
      dispatch(loginAction(user.uid, user.displayName));
      dispatch(disableLoadingAction());
    } catch (error) {
      dispatch(disableLoadingAction());
      swal.fire("Login failed", "Email or Password are invalid!", "error");
    }
  };
};

// Register email and password
export const initRegisterAction = (name, email, password) => {
  return (dispatch) => {
    firebase
      .auth()
      .createUserWithEmailAndPassword(email, password)
      .then(async ({ user }) => {
        await user.updateProfile({
          displayName: name,
          returnSecureToken: true,
        });
        dispatch(loginAction(user.uid, user.displayName));
      })
      .catch((error) => {
        swal.fire("Register failed", error.message, "error");
      });
  };
};

// Register with google sign-in
export const startGoogleLoginAction = () => {
  return (dispatch) => {
    firebase
      .auth()
      .signInWithPopup(googleAuthProvider)
      .then(({ user }) => {
        dispatch(loginAction(user.uid, user.displayName));
      })
      .catch((error) => {
        swal.fire("Google Sign-in Failed!", error.message, "error");
      });
  };
};

// Correct user/password
export const loginAction = (uid, displayName) => ({
  type: types.login,
  payload: { uid, displayName },
});

export const startLogout = () => {
  return async (dispatch) => {
    try {
      await firebase.auth().signOut();
      swal.fire({
        title: "Closing session",
        text: "Wait for it, we're destroying information :)",
        allowOutsideClick: false,
        showCancelButton: false,
        showConfirmButton: false,
      });
      dispatch(logoutAction());
      dispatch(cleanNotesAction());
      window.location.href = "/";
    } catch (error) {
      swal.fire(
        "Logout Failed!",
        "An error occurred while logging out. Please try again. ",
        "error"
      );
    }
  };
};

// Logout Action
export const logoutAction = () => ({
  type: types.logout,
});
