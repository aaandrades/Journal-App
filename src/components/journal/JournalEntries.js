import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { activeNoteAction } from "../../Redux/Actions/notesActions";
import { JournalEntry } from "./JournalEntry";

export const JournalEntries = () => {
  const { notes } = useSelector((state) => state.notes);
  const dispatch = useDispatch();

  const handleClick = (id, note) => {
    dispatch(activeNoteAction(id, note));
  };

  return (
    <div className="journal__entries">
      {notes.map((note) => (
        <JournalEntry key={note.id} {...note} handleClick={handleClick} />
      ))}
    </div>
  );
};
