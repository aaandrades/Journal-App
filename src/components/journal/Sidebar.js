import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { startLogout } from "../../Redux/Actions/authActions";
import { createNoteAction } from "../../Redux/Actions/notesActions";
import { JournalEntries } from "./JournalEntries";

export const Sidebar = () => {
  const dispatch = useDispatch();
  const { name } = useSelector((state) => state.auth);

  const handleLogout = () => {
    dispatch(startLogout());
  };

  const handleCreate = () => {
    dispatch(createNoteAction());
  };

  return (
    <aside className="journal__sidebar">
      <div className="journal__sidebar-navbar">
        <h3>Welcome, {name}!</h3>
      </div>

      <div className="journal__new-entry">
        <div className="journal__new-entry__add" onClick={() => handleCreate()}>
          <i className="far fa-calendar-plus fa-5x mt-5 pointer"></i>
          <p className="pointer">New entry</p>
        </div>
      </div>
      <JournalEntries />
      <button className="logout-btn btn pointer" onClick={() => handleLogout()}>
        Log out
      </button>
    </aside>
  );
};
