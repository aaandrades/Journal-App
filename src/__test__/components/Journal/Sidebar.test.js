/**
 * @jest-environment node
 */
import React from "react";
import { mount } from "enzyme";
import "jsdom-global/register";
import "@testing-library/jest-dom";
import { Provider } from "react-redux";

import { MemoryRouter } from "react-router-dom";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { Sidebar } from "../../../components/journal/Sidebar";
import { startLogout } from "../../../Redux/Actions/authActions";
import { createNoteAction } from "../../../Redux/Actions/notesActions";

jest.mock("../../../Redux/Actions/authActions", () => ({
  startLogout: jest.fn(),
}));

jest.mock("../../../Redux/Actions/notesActions", () => ({
  createNoteAction: jest.fn(),
}));

Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // Deprecated
    removeListener: jest.fn(), // Deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

const initialStore = {
  auth: { uid: "Testing", name: "Andyman" },
  notes: {
    notes: [],
    active: null,
  },
  ui: { loading: false, msgError: null },
};
const middlewares = [thunk];
const mockStore = configureStore(middlewares);

let store = mockStore(initialStore);
store.dispatch = jest.fn();

describe("Sidebar - Component", () => {
  beforeEach(() => {
    store = mockStore(initialStore);
    jest.clearAllMocks();
  });

  const wrapper = mount(
    <Provider store={store}>
      <MemoryRouter>
        <Sidebar />
      </MemoryRouter>
    </Provider>
  );

  test("should match to snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should dispatch startLogout action", () => {
    wrapper.find(".btn").simulate("click");
    expect(startLogout).toHaveBeenCalled();
  });

  test("should dispatch startCreateNewNote action", () => {
    wrapper.find(".journal__new-entry__add").simulate("click");
    expect(createNoteAction).toHaveBeenCalled();
  });
});
