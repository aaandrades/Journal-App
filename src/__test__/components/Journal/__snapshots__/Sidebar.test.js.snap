// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar - Component should match to snapshot 1`] = `
<aside
  className="journal__sidebar"
>
  <div
    className="journal__sidebar-navbar"
  >
    <h3
      className="mt-5"
    >
      <i
        className="far fa-moon"
      />
      <span>
        Andyman
      </span>
    </h3>
    <button
      className="btn"
      onClick={[Function]}
    >
      Log out
    </button>
  </div>
  <div
    className="journal__new-entry"
    onClick={[Function]}
  >
    <i
      className="far fa-calendar-plus fa-5x mt-5 pointer"
    />
    <p
      className="mt-5 pointer"
    >
      New entry
    </p>
  </div>
  <div
    className="journal__entries"
  />
</aside>
`;
